### PR TITLE
Ndev 419 make operator resource list a part of the mem pool

### PR DESCRIPTION
**Pull request by otselnik**

### CODE REVIEW

File: proxy/mempool/mempool_api.py [hunk 1]
```diff
@@ -51,14 +53,16 @@ class MPTxRequest(MPRequest):
...
+    resource: Optional[OperatorResourceInfo] = field(compare=False, default=None)
 
     def __post_init__(self):
         self.gas_price = self.neon_tx.gasPrice
         self.nonce = self.neon_tx.nonce
         self.sender_address = "0x" + self.neon_tx.sender()
         self.type = MPRequestType.SendTransaction
-        tx_hash = self.signature
-        self.log_str = f"MPTxRequest(hash={tx_hash[:10]}..., sender_address=0x{self.sender_address[:10]}..., nonce={self.nonce}, gas_price={self.gas_price})"
+        hash = "0x" + self.neon_tx.hash_signed().hex()
```
Comment (afalaleev) [inline proxy/mempool/mempool_api.py]: What is the reason for this change?
Is there any other place in the source code where the transaction is initialized except memclient::send_raw_transaction?

File: proxy/mempool/mempool_api.py [hunk 2]
```diff
@@ -81,7 +85,9 @@ class MPResultCode(IntEnum):
     Done = 0
     BlockedAccount = 1,
     SolanaUnavailable = 2,
-    Unspecified = 4,
+    PendingTxError = 3,
```
Comment (afalaleev) [inline proxy/mempool/mempool_api.py]: A pending error was removed from the source code.

File: proxy/mempool/operator_resource_list.py [hunk 3]
```diff
@@ -208,7 +124,7 @@ def _validate_operator_balance(self, resource: OperatorResourceInfo) -> None:
                 f'min_operator_balance_to_err = {min_operator_balance_to_err}; '
             )
 
-    def _create_ether_account(self, builder: NeonIxBuilder, resource: OperatorResourceInfo) -> List[NeonTxStage]:
+    def _create_ether_account(self, builder: NeonIxBuilder, resource: OperatorResourceInfo) -> EthereumAddress:
```
Comment (afalaleev) [inline proxy/mempool/operator_resource_list.py]: The function returns List of Stages, not Ethereum address

**Review by afalaleev — CHANGES_REQUESTED** [review]

File: proxy/mempool/mempool_api.py (hunk 2 above)
Comment (otselnik) [reply proxy/mempool/mempool_api.py]: Removed

File: proxy/mempool/operator_resource_list.py (hunk 3 above)
Comment (otselnik) [reply proxy/mempool/operator_resource_list.py]: Fixed

File: proxy/mempool/mempool_api.py (hunk 1 above)
Comment (otselnik) [reply proxy/mempool/mempool_api.py]: Looks like merge artifact - reverted changes.

File: proxy/mempool/mempool.py [hunk 4]
```diff
@@ -20,10 +22,12 @@ class MemPool:
     def __init__(self, executor: IMPExecutor, capacity: int):
         self.info(f"Init mempool schedule with capacity: {capacity}")
         self._tx_schedule = MPTxSchedule(capacity)
+        self._resource_mng = OperatorResourceManager()
         self._schedule_cond = asyncio.Condition()
         self._processing_tasks: List[Tuple[int, asyncio.Task, MPRequest]] = []
         self._process_tx_results_task = asyncio.get_event_loop().create_task(self.check_processing_tasks())
         self._process_schedule_task = asyncio.get_event_loop().create_task(self.process_tx_schedule())
+        self._check_resources_task = asyncio.get_event_loop().create_task(self.check_resources_schedule())
```
Comment (rozhkovdmitrii) [inline proxy/mempool/mempool.py]: Why is it a part of mempool? All the operator management logic should be encapsulated

File: proxy/mempool/operator_resource_list.py [hunk 5]
```diff
@@ -298,5 +215,96 @@ def _unlock_storage_account(self, resource: OperatorResourceInfo, storage_accoun
         cancel_tx_executor.add_blocked_storage_account(storage_info)
         cancel_tx_executor.execute_tx_list()
 
-    def free_resource_info(self, resource: OperatorResourceInfo) -> None:
+
+@logged_group("neon.MemPool")
+class OperatorResourceManager:
+    _free_resource_list = list()
```
Comment (rozhkovdmitrii) [inline proxy/mempool/operator_resource_list.py]: Move it please into `__init__`

File: proxy/mempool/operator_resource_list.py [hunk 6]
```diff
@@ -43,125 +53,42 @@ def public_key(self) -> PublicKey:
     def secret_key(self) -> bytes:
         return self.signer.secret_key()
```
Comment (rozhkovdmitrii) [inline proxy/mempool/operator_resource_list.py:48]: Double blank line

File: proxy/mempool/mempool.py [hunk 7]
```diff
@@ -114,14 +125,22 @@ def _process_mp_result(self, resource_id: int, mp_tx_result: MPTxResult, mp_requ
             log_fn(f"On mp tx result:  {mp_tx_result} - of: {mp_request.log_str}", extra=log_ctx)
 
             if mp_tx_result.code == MPResultCode.BlockedAccount:
+                self._update_locked_resource(mp_request)
```
Comment (rozhkovdmitrii) [inline proxy/mempool/mempool.py]: Isn't it a part of blocked account processing? Why haven't you hidden it behind the `_on_blocked_accounts_result`

File: proxy/mempool/operator_resource_list.py [hunk 8]
```diff
@@ -298,5 +216,96 @@ def _unlock_storage_account(self, resource: OperatorResourceInfo, storage_accoun
...
+        for idx in self._bad_resource_list:
+            self._free_resource_list.append(idx)
+
+        del self._bad_resource_list[:]
+
+    def get_resource(self, tx_hash: str) -> Optional[OperatorResourceInfo]:
+        current_time: int = self._get_current_time()
+        resource: Optional[OperatorResourceInfo] = None
+
+        if self._allocated_resource.get(tx_hash) is not None:
```
Comment (rozhkovdmitrii) [inline proxy/mempool/operator_resource_list.py]:
```python
resource_idx = self._allocated_resource.get(tx_hash)
if resource_ids is not None:
    resource = self._resource_list[resource_idx]
```

File: proxy/mempool/operator_resource_list.py [hunk 9]
```diff
@@ -298,5 +216,96 @@ def _unlock_storage_account(self, resource: OperatorResourceInfo, storage_accoun
...
+            idx = self._allocated_resource[tx_hash]
+            resource = self._resource_list[idx]
```
Comment (rozhkovdmitrii) [inline proxy/mempool/operator_resource_list.py]: `return resource`?

File: proxy/mempool/operator_resource_list.py [hunk 10]
```diff
@@ -1,10 +1,9 @@
-import ctypes
+from __future__ import annotations
+
 import math
-import multiprocessing as mp
-import time
 import traceback
 from datetime import datetime
-from typing import Optional, List, Set
+from typing import Dict, Optional, List
```
Comment (rozhkovdmitrii) [inline proxy/mempool/operator_resource_list.py:8]:
```python
from ..common_neon.environment_data import EVM_LOADER_ID, PERM_ACCOUNT_LIMIT, RECHECK_RESOURCE_LIST_INTERVAL
from ..common_neon.environment_data import MIN_OPERATOR_BALANCE_TO_WARN, MIN_OPERATOR_BALANCE_TO_ERR, STORAGE_SIZE
```
is it possible to get these configuration through the Config?

File: proxy/mempool/operator_resource_list.py [hunk 11]
```diff
@@ -298,5 +216,96 @@ def _unlock_storage_account(self, resource: OperatorResourceInfo, storage_accoun
         cancel_tx_executor.add_blocked_storage_account(storage_info)
         cancel_tx_executor.execute_tx_list()
 
-    def free_resource_info(self, resource: OperatorResourceInfo) -> None:
+
+@logged_group("neon.MemPool")
+class OperatorResourceManager:
+    _free_resource_list = list()
```
Comment (rozhkovdmitrii) [inline proxy/mempool/operator_resource_list.py]: It should be annotated

File: proxy/mempool/mempool.py [hunk 12]
```diff
@@ -173,6 +197,23 @@ async def _kick_tx_schedule(self):
...
+
+    def _free_resource(self, mp_tx_request: MPTxRequest):
+        self._resource_mng.free_resource_info(mp_tx_request.signature, mp_tx_request.resource)
+        mp_tx_request.resource = None
+
+    def _bad_resource(self, mp_tx_request: MPTxRequest):
+        self._resource_mng.bad_resource_info(mp_tx_request.signature, mp_tx_request.resource)
+        mp_tx_request.resource = None
+
+    async def check_resources_schedule(self):
```
Comment (rozhkovdmitrii) [inline proxy/mempool/mempool.py]: It should be hidden behind the OperatorResourceMng

File: proxy/mempool/operator_resource_list.py [hunk 13]
```diff
@@ -298,5 +216,96 @@ def _unlock_storage_account(self, resource: OperatorResourceInfo, storage_accoun
...
+            idx = self._allocated_resource[tx_hash]
+            resource = self._resource_list[idx]
+
+        if len(self._free_resource_list) > 0:
+            idx = self._free_resource_list.pop(0)
+            resource = self._resource_list[idx]
+            self._allocated_resource[tx_hash] = idx
+            self._allocated_resource_access[tx_hash] = current_time
+        else:
+            for blocked_tx_hash, access_time in self._allocated_resource_access:
```
Comment (rozhkovdmitrii) [inline proxy/mempool/operator_resource_list.py:306]: We can't check resource liquidity only through the  `BadResource` error. 
In this case we use the same resource periodically, each time face the same error, each time send resource to the future.

There is another thing you should know, the set of resources can be changed on the fly and operator resource list should be able to solve all the changes.

File: proxy/mempool/operator_resource_list.py [hunk 14]
```diff
@@ -298,5 +216,96 @@ def _unlock_storage_account(self, resource: OperatorResourceInfo, storage_accoun
...
+                idx += 1
+
+        for idx in range(len(self._resource_list)):
+            self._free_resource_list.append(idx)
+
+        if len(self._resource_list) == 0:
+            raise RuntimeError('Operator has NO resources!')
+
+    @staticmethod
+    def recheck_resource_list_interval() -> int:
```
Comment (rozhkovdmitrii) [inline proxy/mempool/operator_resource_list.py]: will gone if you encapsulate all the operator resource management here

File: proxy/mempool/mempool_executor.py [hunk 15]
```diff
@@ -7,14 +7,16 @@
 from typing import Optional, Any
 from neon_py.network import PipePickableDataSrv, IPickableDataServerUser
 
+from proxy.common_neon.errors import BadResourceError
```
Comment (rozhkovdmitrii) [inline proxy/mempool/mempool_executor.py]: relative path

File: proxy/mempool/mempool.py [hunk 16]
```diff
@@ -20,10 +22,12 @@ class MemPool:
     def __init__(self, executor: IMPExecutor, capacity: int):
         self.info(f"Init mempool schedule with capacity: {capacity}")
         self._tx_schedule = MPTxSchedule(capacity)
+        self._resource_mng = OperatorResourceManager()
```
Comment (rozhkovdmitrii) [inline proxy/mempool/mempool.py]: If you encapsulate all the operator management stuff in `OperatorResourceManager` you will have to pass it here by interface to be able test MemPool independently

File: proxy/mempool/mempool.py (hunk 4 above)
Comment (rozhkovdmitrii) [inline proxy/mempool/mempool.py]: Shouldn't be here

File: proxy/mempool/mempool.py [hunk 17]
```diff
@@ -70,6 +74,13 @@ async def process_tx_schedule(self):
                     if mp_request is None:
                         break
 
+                    operator_resource = self._resource_mng.get_resource(mp_request.signature)
+                    if operator_resource is None:
+                        asyncio.get_event_loop().create_task(self._reschedule_tx(mp_request))
```
Comment (rozhkovdmitrii) [inline proxy/mempool/mempool.py]: do not acquire transaction if you don't  have a resource. Change the order please

File: proxy/mempool/mempool.py (hunk 7 above)
Comment (rozhkovdmitrii) [inline proxy/mempool/mempool.py]: If it's a part of blocked account result processing it should be encapsulated there

File: proxy/mempool/mempool.py [hunk 18]
```diff
@@ -114,14 +125,22 @@ def _process_mp_result(self, resource_id: int, mp_tx_result: MPTxResult, mp_requ
             log_fn(f"On mp tx result:  {mp_tx_result} - of: {mp_request.log_str}", extra=log_ctx)
 
             if mp_tx_result.code == MPResultCode.BlockedAccount:
+                self._update_locked_resource(mp_request)
                 self._on_blocked_accounts_result(mp_request, mp_tx_result)
+            elif mp_tx_result.code == MPResultCode.BadResourceError:
+                self._bad_resource(mp_request)
```
Comment (rozhkovdmitrii) [inline proxy/mempool/mempool.py]: If it's a part of bad resource result processing it should be encapsulated there. Here should stay only on call

**Review by rozhkovdmitrii — CHANGES_REQUESTED** [review]

File: proxy/common_neon/errors.py [hunk 19]
```diff
@@ -20,6 +20,11 @@ def __init__(self, *args) -> None:
         RuntimeError.__init__(self, *args)
 
 
+class BadResourceError(Exception):
+    def __int__(self):
```
Comment (rozhkovdmitrii) [inline proxy/common_neon/errors.py]: pass

File: proxy/mempool/mempool.py (hunk 17 above)
Comment (otselnik) [reply proxy/mempool/mempool.py]: To be abble to get same resource for a transaction it need to be acquired first

Comment (rozhkovdmitrii) [reply proxy/mempool/mempool.py]: May be the better way then is to seek for transaction and then, after getting resource, acquire it

**Comment by rozhkovdmitrii** [comment]
OperatorResourceManager should certainly be covered with unittests

File: proxy/common_neon/errors.py (hunk 19 above)
Comment (otselnik) [reply proxy/common_neon/errors.py]: Left it for similarity.

File: proxy/mempool/mempool.py (hunk 16 above)
Comment (otselnik) [reply proxy/mempool/mempool.py]: I think there is no need to hide `OperatorResourceManager` behind interface. It does not interact with anything besides solana accounts on fs.

File: proxy/mempool/mempool.py (hunk 4 above)
Comment (otselnik) [reply proxy/mempool/mempool.py]: Moved to `OperatorResourceManager`

File: proxy/mempool/mempool.py (hunk 17 above)
Comment (otselnik) [reply proxy/mempool/mempool.py]: Implemented

File: proxy/mempool/mempool.py (hunk 7 above)
Comment (otselnik) [reply proxy/mempool/mempool.py]: Fixed.

Comment (otselnik) [reply proxy/mempool/mempool.py]: Fixed.

File: proxy/mempool/mempool.py (hunk 18 above)
Comment (otselnik) [reply proxy/mempool/mempool.py]: Fixed.

File: proxy/mempool/mempool.py (hunk 12 above)
Comment (otselnik) [reply proxy/mempool/mempool.py]: Fixed.

File: proxy/mempool/mempool_executor.py (hunk 15 above)
Comment (otselnik) [reply proxy/mempool/mempool_executor.py]: Fixed.

File: proxy/mempool/operator_resource_list.py (hunk 10 above)
Comment (otselnik) [reply proxy/mempool/operator_resource_list.py:8]: Moved

File: proxy/mempool/operator_resource_list.py (hunk 6 above)
Comment (otselnik) [reply proxy/mempool/operator_resource_list.py:48]: As it should be?

File: proxy/mempool/operator_resource_list.py (hunk 5 above)
Comment (otselnik) [reply proxy/mempool/operator_resource_list.py]: Done

File: proxy/mempool/operator_resource_list.py (hunk 11 above)
Comment (otselnik) [reply proxy/mempool/operator_resource_list.py]: Done

File: proxy/mempool/operator_resource_list.py (hunk 8 above)
Comment (otselnik) [reply proxy/mempool/operator_resource_list.py]: Done

File: proxy/mempool/operator_resource_list.py (hunk 9 above)
Comment (otselnik) [reply proxy/mempool/operator_resource_list.py]: Fixed

File: proxy/common_neon/config.py [hunk 20]
```diff
@@ -26,31 +31,70 @@ def get_pyth_mapping_account(self) -> Optional[str]:
...
+    @abstractmethod
+    def get_recheck_resource_list_interval(self) -> int:
+        """Gets resource recheck interval"""
+
 
 class Config(IConfig):
 
     def get_solana_url(self) -> str:
-        return os.environ.get("SOLANA_URL", "http://localhost:8899")
+        return SOLANA_URL
```
Comment (otselnik) [inline proxy/common_neon/config.py:60]: I changed parameters acquiring. Default values should be defined in one place.

Comment (rozhkovdmitrii) [reply proxy/common_neon/config.py:60]: ok

File: proxy/mempool/mempool.py (hunk 4 above)
Comment (rozhkovdmitrii) [reply proxy/mempool/mempool.py]:
```python
        self._resource_mng.start_periodical_checker()
```

I'm sorry about it, but it's not been. You can safely start it from `OperatorResourceManager` constructor. There is no need to get `MemPool` known about anything inside the `OperatorResourceManager` excluding getting and releasing resource

File: proxy/mempool/operator_resource_list.py (hunk 8 above)
Comment (rozhkovdmitrii) [reply proxy/mempool/operator_resource_list.py]: done

File: proxy/mempool/operator_resource_list.py (hunk 13 above)
Comment (rozhkovdmitrii) [reply proxy/mempool/operator_resource_list.py:306]: @otselnik

File: proxy/mempool/operator_resource_list.py (hunk 6 above)
Comment (rozhkovdmitrii) [reply proxy/mempool/operator_resource_list.py:48]: Yep, these are top level definitions

File: proxy/mempool/mempool.py (hunk 16 above)
Comment (rozhkovdmitrii) [reply proxy/mempool/mempool.py]: If you do it like that you will be able to test MemPool more independently and moreover will be prevented from calling stuff that is not a part of interface like `start_periodical_checker`

File: proxy/mempool/mempool_executor.py (hunk 15 above)
Comment (rozhkovdmitrii) [reply proxy/mempool/mempool_executor.py]: I don't understand why don't you use single line import for these two lines

File: proxy/common_neon/errors.py (hunk 19 above)
Comment (rozhkovdmitrii) [reply proxy/common_neon/errors.py]: I understand it's not critical and I won't pick on it.
But if you do not pass any arguments in it, it's better to use pass to keep it

File: proxy/common_neon/config.py [hunk 21]
```diff
@@ -3,12 +3,17 @@
 from solana.publickey import PublicKey
 import os
```
Comment (rozhkovdmitrii) [inline proxy/common_neon/config.py]: got redundant

File: proxy/common_neon/config.py [hunk 22]
```diff
@@ -26,31 +31,70 @@ def get_pyth_mapping_account(self) -> Optional[str]:
...
+        return MIN_OPERATOR_BALANCE_TO_WARN
+
+    def get_min_operator_balance_to_err(self) -> int:
+        return MIN_OPERATOR_BALANCE_TO_ERR
+
+    def get_perm_account_limit(self) -> int:
+        return PERM_ACCOUNT_LIMIT
+
+    def get_recheck_resource_list_interval(self) -> int:
+        return RECHECK_RESOURCE_LIST_INTERVAL
```
Comment (rozhkovdmitrii) [inline proxy/common_neon/config.py:90]: Extend `__str__` with it please

File: proxy/mempool/operator_resource_list.py [hunk 23]
```diff
@@ -232,23 +143,23 @@ def _create_ether_account(self, builder: NeonIxBuilder, resource: OperatorResour
...
+        for idx, account, stage in zip(range(len(resource.seed_list)), info_list, stage_list):
             if not account:
                 self._make_create_acc_tx(resource, result_stage_list, balance, idx, stage)
                 continue
             elif account.lamports < balance:
                 self._make_refund_tx(builder, resource, refund_stage_list, idx, stage)
                 self._make_create_acc_tx(resource, result_stage_list, balance, idx, stage)
                 continue
-            elif account.owner != PublicKey(EVM_LOADER_ID):
+            elif account.owner != PublicKey(self._config.get_evm_loader_id()):
```
Comment (rozhkovdmitrii) [inline proxy/mempool/operator_resource_list.py]: `self._config.get_evm_loader_id()` - returns public key

File: proxy/mempool/operator_resource_list.py [hunk 24]
```diff
@@ -298,5 +209,112 @@ def _unlock_storage_account(self, resource: OperatorResourceInfo, storage_accoun
...
+
+    @staticmethod
+    def _get_solana_accounts() -> List[SolanaAccount]:
+        return get_solana_accounts()
+
+    @staticmethod
+    def _get_current_time() -> int:
+        return math.ceil(datetime.now().timestamp())
+
+    def _recheck_bad_resource_list(self) -> int:
```
Comment (rozhkovdmitrii) [inline proxy/mempool/operator_resource_list.py]: It doesn't return `int`

File: proxy/mempool/operator_resource_list.py [hunk 25]
```diff
@@ -1,36 +1,46 @@
...
         self.signer = signer
         self.rid = rid
         self.idx = idx
+        self.check_time = 0
         self.ether: Optional[EthereumAddress] = None
         self.storage: Optional[PublicKey] = None
         self.holder: Optional[PublicKey] = None
+        self.seed_list = None
+
+    def fill_storage_holder(self):
```
Comment (rozhkovdmitrii) [inline proxy/mempool/operator_resource_list.py]: The only responsibility of this structure is - to be transferred to the mp executor and be fed to neon transaction processing. I would be better if you left it inside OperatorResourceManager. Please do it

File: proxy/mempool/mempool.py [hunk 26]
```diff
@@ -66,10 +70,22 @@ async def process_tx_schedule(self):
                 await self._schedule_cond.wait_for(self.is_active)
                 self.debug(f"Schedule processing  got awake, condition: {self._schedule_cond.__repr__()}")
                 while self._executor.is_available():
+                    tx_hash = self._tx_schedule.get_tx_good_for_execution()
```
Comment (rozhkovdmitrii) [inline proxy/mempool/mempool.py]: peek_tx_for_execution

File: proxy/mempool/mempool.py [hunk 27]
```diff
@@ -17,9 +19,11 @@ class MemPool:
...
     RESCHEDULE_TIMEOUT_SEC = 0.4
 
-    def __init__(self, executor: IMPExecutor, capacity: int):
-        self.info(f"Init mempool schedule with capacity: {capacity}")
-        self._tx_schedule = MPTxSchedule(capacity)
+    def __init__(self, executor: IMPExecutor, config: IConfig):
+        self.info(f"Init mempool schedule with capacity: {config.get_mempool_capacity()}")
+        self._tx_schedule = MPTxSchedule(config.get_mempool_capacity())
+        self._resource_mng = OperatorResourceManager(config)
+        self._resource_mng.start_periodical_checker()
```
Comment (rozhkovdmitrii) [inline proxy/mempool/mempool.py]: can be isolated inside the manager

File: proxy/mempool/operator_resource_list.py [hunk 28]
```diff
@@ -298,5 +209,112 @@ def _unlock_storage_account(self, resource: OperatorResourceInfo, storage_accoun
...
+        self._bad_resource_list.append(resource.idx)
+
+    def free_resource_info(self, tx_hash: str, resource: OperatorResourceInfo) -> None:
+        current_time: int = self._get_current_time()
+        del self._allocated_resource[tx_hash]
+        del self._allocated_resource_access[tx_hash]
+        self._resource_list[resource.idx].check_time = current_time
         self._free_resource_list.append(resource.idx)
+
+    def deallocate_resource(self, tx_hash: str, resource: OperatorResourceInfo) -> None:
```
Comment (rozhkovdmitrii) [inline proxy/mempool/operator_resource_list.py]: 1. release_resource
2. why do you use resource parameter, all the needed information is already inside. It looks like if you get rid of it, the logic will straighten up

File: proxy/mempool/mempool.py [hunk 29]
```diff
@@ -130,40 +148,52 @@ def _process_mp_result(self, resource_id: int, mp_tx_result: MPTxResult, mp_requ
             self._executor.release_resource(resource_id)
 
     def _on_blocked_accounts_result(self, mp_tx_request: MPTxRequest, mp_tx_result: MPTxResult):
+        self._update_locked_resource(mp_tx_request)
```
Comment (rozhkovdmitrii) [inline proxy/mempool/mempool.py:151]: it's workaround should gone in the next PR as we got agreed

File: proxy/mempool/mempool.py [hunk 30]
```diff
@@ -130,40 +148,52 @@ def _process_mp_result(self, resource_id: int, mp_tx_result: MPTxResult, mp_requ
...
         self.debug(f"For tx: {mp_tx_request.log_str} - got blocked account transaction status: {mp_tx_result.data}. "
                    f"Will be rescheduled in: {self.RESCHEDULE_TIMEOUT_SEC} sec.")
         asyncio.get_event_loop().create_task(self._reschedule_tx(mp_tx_request))
 
+    def _on_bad_resource_error_result(self, mp_tx_request: MPTxRequest, mp_tx_result: MPTxResult):
+        self._bad_resource(mp_tx_request)
```
Comment (rozhkovdmitrii) [inline proxy/mempool/mempool.py]: We already are inside "bad resource" handler )

File: proxy/mempool/mempool.py [hunk 31]
```diff
@@ -130,40 +148,52 @@ def _process_mp_result(self, resource_id: int, mp_tx_result: MPTxResult, mp_requ
             self._executor.release_resource(resource_id)
 
     def _on_blocked_accounts_result(self, mp_tx_request: MPTxRequest, mp_tx_result: MPTxResult):
+        self._update_locked_resource(mp_tx_request)
         self.debug(f"For tx: {mp_tx_request.log_str} - got blocked account transaction status: {mp_tx_result.data}. "
                    f"Will be rescheduled in: {self.RESCHEDULE_TIMEOUT_SEC} sec.")
         asyncio.get_event_loop().create_task(self._reschedule_tx(mp_tx_request))
 
+    def _on_bad_resource_error_result(self, mp_tx_request: MPTxRequest, mp_tx_result: MPTxResult):
```
Comment (rozhkovdmitrii) [inline proxy/mempool/mempool.py:156]: `on_bad_resource_result`

File: proxy/mempool/mempool.py [hunk 32]
```diff
@@ -130,40 +148,52 @@ def _process_mp_result(self, resource_id: int, mp_tx_result: MPTxResult, mp_requ
...
+        self._bad_resource(mp_tx_request)
+        self.debug(f"For tx: {mp_tx_request.log_str} - got bad resource error status: {mp_tx_result.data}. "
+                   f"Will be rescheduled in: {self.RESCHEDULE_TIMEOUT_SEC} sec.")
+        asyncio.get_event_loop().create_task(self._reschedule_tx(mp_tx_request))
```
Comment (rozhkovdmitrii) [inline proxy/mempool/mempool.py:160]: It seems logging and rescheduling is a separate thing.
The first action to be logged is the "bad resource" result

The second is rescheduling (with logging), now it seems is able to be extracted

File: proxy/mempool/mempool.py [hunk 33]
```diff
@@ -130,40 +148,52 @@ def _process_mp_result(self, resource_id: int, mp_tx_result: MPTxResult, mp_requ
...
+
     def _on_solana_unavailable_result(self, mp_tx_request: MPTxRequest, mp_tx_result: MPTxResult):
+        self._update_locked_resource(mp_tx_request)
         self.debug(f"For tx: {mp_tx_request.log_str} - got solana unavailable status: {mp_tx_result.data}. "
                    f"Will be rescheduled in: {self.RESCHEDULE_TIMEOUT_SEC} sec.")
         asyncio.get_event_loop().create_task(self._reschedule_tx(mp_tx_request))
 
     def _on_low_gas_price_result(self, mp_tx_request: MPTxRequest, mp_tx_result: MPTxResult):
+        self._free_resource(mp_tx_request)
```
Comment (rozhkovdmitrii) [inline proxy/mempool/mempool.py:169]: `release_resource(mp_tx_request.hash)`

File: proxy/mempool/mempool.py [hunk 34]
```diff
@@ -130,40 +148,52 @@ def _process_mp_result(self, resource_id: int, mp_tx_result: MPTxResult, mp_requ
...
         asyncio.get_event_loop().create_task(self._reschedule_tx(mp_tx_request))
 
+    def _on_bad_resource_error_result(self, mp_tx_request: MPTxRequest, mp_tx_result: MPTxResult):
+        self._bad_resource(mp_tx_request)
+        self.debug(f"For tx: {mp_tx_request.log_str} - got bad resource error status: {mp_tx_result.data}. "
+                   f"Will be rescheduled in: {self.RESCHEDULE_TIMEOUT_SEC} sec.")
+        asyncio.get_event_loop().create_task(self._reschedule_tx(mp_tx_request))
+
     def _on_solana_unavailable_result(self, mp_tx_request: MPTxRequest, mp_tx_result: MPTxResult):
+        self._update_locked_resource(mp_tx_request)
```
Comment (rozhkovdmitrii) [inline proxy/mempool/mempool.py:163]: it's workaround should gone in the next PR as we got agreed

File: proxy/mempool/mempool.py [hunk 35]
```diff
@@ -173,6 +203,18 @@ async def _kick_tx_schedule(self):
     def on_resource_got_available(self, resource_id: int):
         asyncio.get_event_loop().create_task(self._kick_tx_schedule())
 
+    def _update_locked_resource(self, mp_tx_request: MPTxRequest):
+        self._resource_mng.update_allocated_resource(mp_tx_request.signature, mp_tx_request.resource)
+        mp_tx_request.resource = None
```
Comment (rozhkovdmitrii) [inline proxy/mempool/mempool.py]: It's certainly shouldn't be here
It looks like an "ill sign"
Resource shouldn't be a part of request, it should be passed aside of it as it is or as a part of "processing context", we discussed it earlier
All the information you need is the transaction hash to release and get resource

File: proxy/mempool/mempool_executor.py [hunk 36]
```diff
@@ -76,17 +80,17 @@ def execute_neon_tx(self, mp_tx_request: MPTxRequest):
     def execute_neon_tx_impl(self, mp_tx_request: MPTxRequest):
         neon_tx = mp_tx_request.neon_tx
         neon_tx_exec_cfg = mp_tx_request.neon_tx_exec_cfg
+        resource = mp_tx_request.resource
+
+        if not ResourceInitializer(self._config, self._solana_interactor).init_resource(mp_tx_request.resource):
```
Comment (rozhkovdmitrii) [inline proxy/mempool/mempool_executor.py]: Don't understand should it be inside the OperatorResourceManager processing or not

File: proxy/mempool/mempool.py [hunk 37]
```diff
@@ -173,6 +203,18 @@ async def _kick_tx_schedule(self):
     def on_resource_got_available(self, resource_id: int):
         asyncio.get_event_loop().create_task(self._kick_tx_schedule())
 
+    def _update_locked_resource(self, mp_tx_request: MPTxRequest):
+        self._resource_mng.update_allocated_resource(mp_tx_request.signature, mp_tx_request.resource)
```
Comment (rozhkovdmitrii) [inline proxy/mempool/mempool.py]: self._resource_mng.update_allocated_resource(mp_tx_request.signature~~, mp_tx_request.resource~~)

File: proxy/mempool/mempool_executor.py [hunk 38]
```diff
@@ -76,17 +80,17 @@ def execute_neon_tx(self, mp_tx_request: MPTxRequest):
     def execute_neon_tx_impl(self, mp_tx_request: MPTxRequest):
```
Comment (rozhkovdmitrii) [inline proxy/mempool/mempool_executor.py]: `def execute_neon_tx_impl(self, mp_tx_request: MPTxRequest, mp_exec_ctx: MPExecCtx):` 

resource is a context like an emulating result derivatives, please do not  stuff all things into `MPTxRequest`. 
MpTxRequest is not good and it's my fault, please don't make it worse

File: proxy/mempool/mempool_schedule.py [hunk 39]
```diff
@@ -87,6 +87,11 @@ def add_tx(self, mp_tx_request: MPTxRequest) -> MPSendTxResult:
     def get_tx(self):
         return None if self.is_empty() else self._tx_list[0]
 
+    def get_tx_good_for_processing(self):
+        if self.is_processing():
+            return None
+        return self.get_tx()
```
Comment (rozhkovdmitrii) [inline proxy/mempool/mempool_schedule.py]: `self.peek_tx()`

File: proxy/mempool/mempool_schedule.py [hunk 40]
```diff
@@ -249,6 +253,20 @@ def acquire_tx_for_execution(self) -> Optional[MPTxRequest]:
 
         return tx
 
+    def get_tx_good_for_execution(self) -> Optional[str]:
```
Comment (rozhkovdmitrii) [inline proxy/mempool/mempool_schedule.py]: Here are two methods get_tx_good_for_processing and this one )))

File: proxy/common_neon/errors.py (hunk 19 above)
Comment (rozhkovdmitrii) [reply proxy/common_neon/errors.py]: done

File: proxy/mempool/mempool.py (hunk 37 above)
Comment (rozhkovdmitrii) [reply proxy/mempool/mempool.py]: done

File: proxy/mempool/mempool_executor.py (hunk 38 above)
Comment (rozhkovdmitrii) [reply proxy/mempool/mempool_executor.py]: done in a little bit different way

File: proxy/mempool/mempool_executor.py (hunk 36 above)
Comment (rozhkovdmitrii) [reply proxy/mempool/mempool_executor.py]: Agree it's a task related to bringing checking balances and keys on the separate process

**Review by rozhkovdmitrii — APPROVED** [review]

**Review by afalaleev — APPROVED** [review]

**Pull request merged**